### PR TITLE
Implement `_geoBoundingBox` filter

### PR DESCRIPTION
**Pull request by gmourier**
# Pull Request

## Related issue
Fixes #[2761](https://github.com/meilisearch/meilisearch/issues/2761)

## What does this PR do?
- Implements `_geoBoundingBox((lat, lng), (lat, lng))` filter

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: milli/src/search/facet/filter.rs [hunk 1]
```diff
@@ -423,6 +423,104 @@ impl<'a> Filter<'a> {
...
+                    )?;
+
+                    let geo_lng_token =
+                        Token::new(top_left_point[1].span, Some("_geo.lng".to_string()));
+                    let min_lng_token =
+                        Token::new(top_left_point[1].span, Some("-180.0".to_string()));
+                    let max_lng_token =
+                        Token::new(top_left_point[1].span, Some("180.0".to_string()));
+
+                    let selected_lng = if top_left[1] > bottom_right[1] {
```
Comment (loiclec) [inline milli/src/search/facet/filter.rs:473]: nice! That's mostly what I was worried about :)

File: filter-parser/src/error.rs [hunk 2]
```diff
@@ -141,23 +143,29 @@ impl<'a> Display for Error<'a> {
...
             }
             ErrorKind::ExpectedEof => {
                 writeln!(f, "Found unexpected characters at the end of the filter: `{}`. You probably forgot an `OR` or an `AND` rule.", escaped_input)?
             }
-            ErrorKind::Geo => {
+            ErrorKind::GeoRadius => {
                 writeln!(f, "The `_geoRadius` filter expects three arguments: `_geoRadius(latitude, longitude, radius)`.")?
             }
+            ErrorKind::GeoBoundingBox => {
+                writeln!(f, "The `_geoBoundingBox` filter expects two pair of arguments: `_geoBoundingBox((latitude, longitude), (latitude, longitude))`.")?
```
Comment (loiclec) [inline filter-parser/src/error.rs:158]: typo: `two pairS`

File: filter-parser/src/value.rs [hunk 3]
```diff
@@ -91,11 +94,31 @@ pub fn parse_value<'a>(input: Span<'a>) -> IResult<Token<'a>> {
...
+
+    match parse_geo_bounding_box(input) {
+        Ok(_) => {
+            return Err(nom::Err::Failure(Error::new_from_kind(
+                input,
+                ErrorKind::MisusedGeoBoundingBox,
+            )))
+        }
+        // if we encountered a failure it means the user badly wrote a _geoBoundingBox filter.
+        // But instead of showing him how to fix his syntax we are going to tell him he should not use this filter as a value.
```
Comment (loiclec) [inline filter-parser/src/value.rs:116]: nitpick: I'd prefer to use gender-neutral pronouns in examples
`he/him` -> `they/them`

**Review by loiclec — APPROVED** [review]
Great PR, thank you :)
I was a bit worried about some implementation details when crossing the line of 180deg longitude and about some validation logic, but it looks like everything is handled properly 😄

File: filter-parser/src/value.rs (hunk 3 above)
Comment (gmourier) [reply filter-parser/src/value.rs:116]: It's a comment already existing on other `parse_*` methods related to the geosearch. See line `101`. I will update them accordingly.

File: filter-parser/src/lib.rs [hunk 4]
```diff
@@ -316,6 +318,38 @@ fn parse_geo_radius(input: Span) -> IResult<FilterCondition> {
...
+    )(input)
+    .map_err(|e| e.map(|_| Error::new_from_kind(input, ErrorKind::GeoBoundingBox)));
+
+    let (input, args) = parsed?;
+
+    if args.len() != 2 || args[0].len() != 2 || args[1].len() != 2 {
+        return Err(nom::Err::Failure(Error::new_from_kind(input, ErrorKind::GeoBoundingBox)));
+    }
+
+    //TODO: Check sub array length
```
[the remark this answers is not in the document]
Comment (gmourier) [reply filter-parser/src/lib.rs:345]: Yep :exploding_head:

**Comment by curquiza** [comment]
@gmourier sorry, we took time to prioritize the review of your PR. Could you rebase it please?

File: milli/src/search/facet/filter.rs [hunk 5]
```diff
@@ -433,6 +433,104 @@ impl<'a> Filter<'a> {
                     }))?
                 }
             }
+            FilterCondition::GeoBoundingBox { top_left_point, bottom_right_point } => {
+                if filterable_fields.contains("_geo") {
+                    let top_left: [f64; 2] =
+                        [top_left_point[0].parse()?, top_left_point[1].parse()?];
```
Comment (gmourier) [inline milli/src/search/facet/filter.rs:439]:
```suggestion
                        [top_left_point[0].parse_finite_float()?, top_left_point[1].parse_finite_float()?];
```

File: milli/src/search/facet/filter.rs [hunk 6]
```diff
@@ -433,6 +433,104 @@ impl<'a> Filter<'a> {
...
+                    let bottom_right: [f64; 2] =
+                        [bottom_right_point[0].parse()?, bottom_right_point[1].parse()?];
```
Comment (gmourier) [inline milli/src/search/facet/filter.rs:441]:
```suggestion
                        [bottom_right_point[0].parse_finite_float()?, bottom_right_point[1].parse_finite_float()?];
```

**Comment by curquiza** [comment]
bors try

**Comment by curquiza** [comment]
Closed in favor of https://github.com/meilisearch/meilisearch/pull/3405

**Pull request closed without merging**